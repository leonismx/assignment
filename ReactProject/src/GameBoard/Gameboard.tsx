import React from 'react';
import { FlatList } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { CARD_PAIRS_VALUE } from '../App';
import CardComponent from '../Card/Card';
import {
  openCard,
  selectCards,
  clearCards,
  disableAllCards,
  selectIfCardsShouldBeDisabled,
  enableAllCards,
  Card,
} from '../slices/gameSlice';
import { GAMEBOARD_MARGIN, getColumns } from '../Utils/dimension';
interface RenderProps {
  item: Card;
  index: number;
}

const RenderItem = (props: RenderProps) => {
  const { item, index } = props;
  const dispatch = useDispatch();

  const triggerCardTapped = (index: number) => {
    dispatch(openCard(index));
    dispatch(enableAllCards());
  };

  const triggerClearFlippedCard = (index: number) => {
    dispatch(clearCards(index));
    dispatch(enableAllCards());
  };
  const disable = () => {
    dispatch(disableAllCards());
  };

  return (
    <CardComponent
      cardIndex={index}
      card={item}
      cardTapped={triggerCardTapped}
      clearFlippedCards={triggerClearFlippedCard}
      disable={useSelector(selectIfCardsShouldBeDisabled)}
      disableAllCards={disable}
    />
  );
};

const GameBoard = () => {
  const cards = useSelector(selectCards);
  const col = getColumns(CARD_PAIRS_VALUE);
  return (
    <FlatList
      removeClippedSubviews={true}
      style={{ margin: GAMEBOARD_MARGIN }}
      data={cards}
      keyExtractor={(item) => item.id.toString()}
      numColumns={col}
      renderItem={(props) => <RenderItem {...props} />}
    />
  );
};

export default GameBoard;
