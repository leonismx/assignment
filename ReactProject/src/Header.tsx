import React from 'react';
import { Button, View, Text, StyleSheet } from 'react-native';
import { useSelector } from 'react-redux';
import { selectStepsTaken } from './slices/gameSlice';

interface HeaderProps {
  restartBtnTapped: () => void;
}

export const StepsComponent = () => {
  const steps = useSelector(selectStepsTaken);

  return (
    <Text style={styles.stepsTextStyle}>
      STEPS:<Text style={styles.stepsNumberStyle}> {steps}</Text>
    </Text>
  );
};

const Header = (props: HeaderProps) => {
  return (
    <View style={styles.headerContainer}>
      <Button title="Restart" onPress={props.restartBtnTapped}></Button>
      <StepsComponent />
    </View>
  );
};

const styles = StyleSheet.create({
  stepsTextStyle: {
    fontSize: 20,
    textAlignVertical: 'center',
  },
  stepsNumberStyle: {
    color: 'skyblue',
    fontSize: 25,
    fontWeight: 'bold',
    textAlignVertical: 'center',
  },
  headerContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    margin: 15,
  },
});

export default Header;
