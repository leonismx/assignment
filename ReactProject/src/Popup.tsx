import { useEffect } from 'react';
import { Alert } from 'react-native';
import { useSelector } from 'react-redux';
import { selectIfGameEnded } from './slices/gameSlice';

export interface PopupProps {
  callBack: () => void;
}

const Popup = (props: PopupProps) => {
  const gameCompleted = useSelector(selectIfGameEnded);

  const showAlert = () =>
    Alert.alert('Congratulations!', 'You have won!', [
      {
        text: 'OK',
        onPress: () => {
          props.callBack();
        },
        style: 'cancel',
      },
    ]);

  useEffect(() => {
    if (gameCompleted) {
      //set a delay here for better ux
      setTimeout(() => {
        showAlert();
      }, 600);
    }
  }, [gameCompleted]);

  return null;
};

export default Popup;
