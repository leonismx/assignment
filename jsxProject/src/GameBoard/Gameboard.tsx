import React, { useMemo } from 'react';
import { FlatList } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { CARD_PAIRS_VALUE } from '../App';
import CardComponent from '../Card/Card';
import {
  openCard,
  selectCards,
  clearCards,
  disableAllCards,
  selectIfCardsShouldBeDisabled,
  enableAllCards,
  Card,
} from '../slices/gameSlice';
import { GAMEBOARD_MARGIN, getColumns } from '../Utils/dimension';
interface RenderProps {
  item: Card;
  index: number;
}

const RenderItem = (props: RenderProps) => {
  const { item, index } = props;
  const dispatch = useDispatch();

  const triggerCardTapped = (index: number) => {
    dispatch(openCard(index));
    dispatch(enableAllCards());
  };

  const triggerClearFlippedCard = (index: number) => {
    dispatch(clearCards(index));
    dispatch(enableAllCards());
  };
  const disable = () => {
    dispatch(disableAllCards());
  };

  return (
    <CardComponent
      cardIndex={index}
      card={item}
      opened={item.opened}
      restart={item.restart}
      cardTapped={triggerCardTapped}
      reset={item.reset}
      clearFlippedCards={triggerClearFlippedCard}
      disable={useSelector(selectIfCardsShouldBeDisabled)}
      disableAllCards={disable}
    />
  );
};

const GameBoard = () => {
  const cards = useSelector(selectCards);
  const col = getColumns();

  return (
    <FlatList
      removeClippedSubviews={true}
      style={{ margin: GAMEBOARD_MARGIN }}
      data={cards}
      keyExtractor={(item) => item.id.toString()}
      numColumns={col}
      renderItem={(props) => <RenderItem {...props} />}
    />
  );
};

export default GameBoard;
