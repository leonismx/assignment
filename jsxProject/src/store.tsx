import { configureStore } from '@reduxjs/toolkit';
import gameReducer from './slices/gameSlice';
export default configureStore({
  reducer: {
    gameboard: gameReducer,
  },
});
