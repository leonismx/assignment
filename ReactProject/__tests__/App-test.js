/**
 * @format
 */

import Header from '../src/Header';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import store from '../src/store';
import { Provider } from 'react-redux';
import GameBoard from '../src/GameBoard/Gameboard';
import Popup from '../src/Popup';

test('renders correctly Header component', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Header />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly Gameboard component', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <GameBoard />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly Pop up component', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Popup />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
