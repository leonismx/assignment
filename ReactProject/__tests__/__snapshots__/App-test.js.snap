// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Gameboard component 1`] = `
<RCTScrollView
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "margin": 15,
    }
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View />
</RCTScrollView>
`;

exports[`renders correctly Header component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "margin": 15,
    }
  }
>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Restart
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "fontSize": 20,
        "textAlignVertical": "center",
      }
    }
  >
    STEPS:
    <Text
      style={
        Object {
          "color": "skyblue",
          "fontSize": 25,
          "fontWeight": "bold",
          "textAlignVertical": "center",
        }
      }
    >
       
      0
    </Text>
  </Text>
</View>
`;

exports[`renders correctly Pop up component 1`] = `null`;
